properties/base/base.py

```diff
@@ -510,36 +514,52 @@ def deserialize(cls, value, trusted=False, strict=False,                   #pyli
         if not isinstance(value, dict):
             raise ValueError('HasProperties must deserialize from dictionary')
-        input_class = value.get('__class__')
-        if not input_class or input_class == cls.__name__:
-            pass
-        elif trusted and input_class in cls._REGISTRY:
-            cls = cls._REGISTRY[input_class]
-        elif strict:
-            raise utils.ValidationError(
-                'Class name {} from input dictionary does not match input '
-                'class {}'.format(input_class, cls.__name__))
-        kwargs.update({'trusted': trusted, 'strict': strict})
+        cls = cls._deserialize_class(value.get('__class__'), trusted, strict)
+        instance = kwargs.pop('_instance', None)
+        if instance is not None and not isinstance(instance, cls):
+            raise ValueError(
+                'Input _instance must be of class {}, not {}'.format(
+                    cls.__name__, instance.__class__.__name__
+                )
+            )
         state, unused = utils.filter_props(cls, value, True)
         unused.pop('__class__', None)
         if unused and strict:
-            raise utils.ValidationError(
+            raise ValueError(
                 'Unused properties during deserialization: {}'.format(
                     ', '.join(unused)
                 )
             )
+        kwargs.update({'trusted': trusted, 'strict': strict})
         newstate = {}
         for key, val in iteritems(state):
             newstate[key] = cls._props[key].deserialize(val, **kwargs)
         mutable, immutable = utils.filter_props(cls, newstate, False)
         with handlers.listeners_disabled():
-            newinst = cls(**mutable)
+            if instance is None:
+                instance = cls(**mutable)
+            else:
+                for key, val in iteritems(mutable):
+                    setattr(instance, key, val)
         for key, val in iteritems(immutable):
-            valid_val = cls._props[key].validate(newinst, val)
-            newinst._backend[key] = valid_val
-        if assert_valid and not newinst.validate():
+            valid_val = cls._props[key].validate(instance, val)
+            instance._backend[key] = valid_val
+        if assert_valid and not instance.validate():
             raise utils.ValidationError('Deserialized instance is not valid')
-        return newinst
+        return instance
+
+    @classmethod
+    def _deserialize_class(cls, input_class, trusted=False, strict=False):
+        """Returns the HasProperties class to use for deserialization"""
+        if not input_class or input_class == cls.__name__:
+            pass
+        elif trusted and input_class in cls._REGISTRY:
+            cls = cls._REGISTRY[input_class]
+        elif strict:
+            raise ValueError(
+                'Class name {} from deserialization input dictionary does '
+                'not match input class {}'.format(input_class, cls.__name__))
+        return cls
```

Comment:
Could this be simplified a little? I'm not sure exactly how, but the boolean logic is pretty complicated. I wonder if it's possible to exit early on the trusted flag:

```python
if (not trusted and input_class != cls.__name__) or input_class not in cls._REGISTRY:
    if not strict:
        return cls
    raise ValueError(
        'Class name {} from deserialization input dictionary does '
        'not match input class {}'.format(input_class, cls.__name__))
return cls._REGISTRY[input_class]
```

I don't know, maybe that's not better. Just having an if statement with a `pass` in it feels weird. Alternatively, maybe just a `return cls` in place of `pass`?

Reply:
Hm, yeah, I'm not too particular.

Reply:
Ok, these aren't quite identical if `input_class` is `None`. I think I'll leave my `if` structure, but be more generous with `return`s.

Reply:
Cool